miprometheus/utils/sampler_factory.py

```diff
@@ -0,0 +1,169 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+sampler_factory.py: Factory building samplers used by PyTorch's DataLoaders.
+
+"""
+__author__ = "Tomasz Kornuta"
+
+import os
+import logging
+import torch.utils.data.sampler
+
+
+class SamplerFactory(object):
+    """
+    Class returning sampler depending on the name provided in the \
+    list of parameters.
+    """
+
+    @staticmethod
+    def build(problem, params):
+        """
+        Static method returning particular sampler, depending on the name \
+        provided in the list of parameters.
+
+        :param problem: Instance of an object derived from the Problem class.
+        :type problem: ``problems.Problem``
+
+        :param params: Parameters used to instantiate the sampler.
+        :type params: ``utils.param_interface.ParamInterface``
+
+        ..note::
+
+            ``params`` should contains the exact (case-sensitive) class name of the sampler to instantiate.
+
+        :return: Instance of a given sampler or None if section not present (None) or coudn't build the sampler.
+        """
+        logging.basicConfig(level=logging.INFO)
+        logger = logging.getLogger('SamplerFactory')
+
+        # Check if sampler is required, i.e. 'sampler' section is empty.
+        if not params:
+            return None
+
+        try: 
+            # Check presence of the name attribute.
+            if 'name' not in params:
+                raise Exception("Sampler configuration section does not contain the key 'name'")
+
+            # Get the class name.
+            name = params['name']
+
+            # Verify that the specified class is in the controller package.
+            if name not in dir(torch.utils.data.sampler):
+                raise Exception("Could not find the specified class '{}' in the controllers package".format(name))
+
+            # Get the actual class.
+            sampler_class = getattr(torch.utils.data.sampler, name)
+
+            # Ok, proceed.
+            logger.info('Loading the {} controller from {}'.format(name, sampler_class.__module__))
+
+            # Handle "special" case.
+            if sampler_class.__name__ == 'SubsetRandomSampler':
+
+                indices = params['indices']
+
+                # Analyze the type.
+                if type(indices) == str:
+
+                    # Try to open the file.
+                    try:
+                        # from expanduser()'s doc: If the expansion fails or if the path does not begin
+                        # with a tilde, the path is returned unchanged. -> So operation below should be safe.
+                        file = open(os.path.expanduser(indices), "r")
```

Comment:
The `os.path.expanduser()` function is pretty safe, and handles very well the case if the path does not start with a tilde. From the [doc](https://docs.python.org/3.7/library/os.path.html#os.path.expanduser):

> If the expansion fails or if the path does not begin with a tilde, the path is returned unchanged.

So we can apply it directly without making any `try .. except` block:

![expanduser](https://user-images.githubusercontent.com/33840936/47575867-e7aa0080-d8f7-11e8-9c15-7ec962c8f5ea.png)

I did the change.